### DISCUSSION TITLE
Timestamp issues at the beginning of files when using Whisper Annotations

### DISCUSSION THREAD

**Discussion by scott-vector** [lhotse-speech/lhotse#868] [Q&A]
This is a great effort, all at lhotse-speech, so thank you.

When using this to get timestamps from audio with beeps/other noises, the first portion of the file's timestamps do not match up. It begins to sync up around 30-40s in. I've tried to use just the speech, but it still takes the beeps into account. Any suggestions would be great. Thanks.

**Comment by pzelasko** [comment]
Depending on how noisy your audio is, it might just be difficult to correctly transcribe for the model. Make sure you're using the latest version of Lhotse, as we fixed some timestamp related issues recently. Also if you're using the default settings, you can try `medium.en` or `large` version of Whisper for better results.

**Reply by scott-vector** [reply]
I've tried all of the models, including large. It still spits out a bundle of segments with bad stamps at the beginning of the file. It's confusing things like beeps with speech and thinking it needs to set up a supervision (or something). When I use VAD, it includes this portion as a segment as well.

**Reply by pzelasko** [reply]
> I've tried all of the models, including large. It still spits out a bundle of segments with bad stamps at the beginning of the file. It's confusing things like beeps with speech and thinking it needs to set up a supervision (or something). 

It looks like it's just how the model transcribes, can't do too much about it.

> When I use VAD, it includes this portion as a segment as well.

How so? You can use VAD timestamps to truncate the supervisions (post processing) or to give smaller speech segments to the model to transcribe (pre processing, the better option).

**Reply by scott-vector** [reply]
Whisper picks up on the words in the audio properly, it just has the timestamps wrong. Is this is a lhotse thing, as Whisper simply delivers the words to lhotse, which then does the forced alignment?

**Reply by desh2608** [reply]
Lhotse does not do any forced alignment AFAIK. Whisper provides approximate time-stamps for the transcribed segments, and the Lhotse workflow simply uses that to create the supervisions. 

Whisper is based on label-synchronous decoding, so you cannot expect it to produce good time-stamps. In fact, the timestamps it produces are at the segment level instead of at frame-level.

**Reply by scott-vector** [reply]
@desh2608 Thank you for that explanation. If Whisper only has segment level timestamps, where does Lhotse get the "alignments" for each word?

**Reply by desh2608** [reply]
I don't think the [Whisper workflow](https://github.com/lhotse-speech/lhotse/blob/master/lhotse/workflows/whisper.py) adds any word-level alignments.

**Reply by pzelasko** [reply]
If you run the torchaudio wav2vec2 workflow, it uses a wav2vec2 ASR to perform forced alignment given the transcripts in the supervision

**Reply by scott-vector** [reply]
My WAV file is a phone call with a few rings at the beginning. The first 10-15 words get bad timestamps (they are listed as happening while the phone ringing is happening), then once it gets into the actual speech, the timestamps become aligned for the subsequent words.

This is what I am doing, which I think equates to what @pzelasko mentioned as far as the wav2vec2 workflow. If I am missing a step, let me know. The reading at the end is to give me the clean JSON list for the alignments themselves. I am hoping there is a better way to get this information as well.

```
recordings = RecordingSet.from_dir(
    "wav_temp", pattern=f'{args.input_file}')
cuts = annotate_with_whisper(recordings, "English", "large", "cuda")
cuts_aligned = align_with_torchaudio(cuts)

with CutSet.open_writer(f'wav_temp/{rnd}/cuts.jsonl.gz') as writer:
    for cut in cuts_aligned:
        writer.write(cut)

cuts = CutSet.from_file(
    f'wav_temp/{rnd}/cuts.jsonl.gz'
)

listObj = []

decomp = cuts.decompose(output_dir=f'wav_temp/{rnd}')

with gzip.open(f'wav_temp/{rnd}/supervisions.jsonl.gz', 'rt') as f:
    for line in f:
        listObj.append(line)

json_object = json.dumps(listObj)
print(json_object)
```

**Reply by desh2608** [reply]
Ohh, you are using the forced alignment workflow. I think the default model used there is `WAV2VEC2_ASR_BASE_960H` which is probably messing up the alignment in the beginning because it is trained only on LibriSpeech (which has no telephone ring noise). You could try to use one of the other model by specifying it as `bundle_name` in the `align_with_torchaudio` function, but I don't think any of the pretrained torchaudio pipelines have been trained with background noise. Another option would be to truncate the start of the audio to the start of the first segment as identified by Whisper, and then do the forced alignment on the truncated cut.

**Reply by pzelasko** [reply]
I think the torchaudio workflow already uses the supervision timestamps. It looks like the two best bets are:
- use a model for telephone domain for forced alignment (not sure if a pretrained one is available)
- insert an extra processing step with VAD to cut out the nonspeech region in supervision segments (basically truncate each supervision's start/duration)

BTW the decompose thing is not necessary, you can do:
```python
for cut in cuts_aligned:
  for sup in cut.supervisions:
    print(cut.id, sup.id, sup.alignment["word"])
```

or if you really want json objects: `for d in cuts_aligned.to_dicts(): print(d)`

**Reply by scott-vector** [reply]
@pzelasko thank you for the guidance on reducing the decompose, that helped testing a LOT.

I tried several pipelines, and they all have the timestamps starting at the exact same time in the recording (about 1.2s into it). This is during the ringing, while the speech doesn't actually start until around 15s. In the 5-6 I tested, I received precisely the same offset. 

@desh2608, the first segment Whisper identifies covers the entire range (0-22.4s), so I am not sure what it thinks is right or wrong. VAD has issues as well. I've tried a number of different VAD solutions, including silero, and none I've found so far are able to differentiate between the phone's ring and a human speaking (meaning it counts the rings as spoken portions of the call).

**Reply by pzelasko** [reply]
Probably you need to train/fine-tune your own VAD on this kind of data to make it work...

**Reply by scott-vector** [reply]
@pzelasko Yes, this is where my mind was headed... custom VAD and probably a telephone-based pipeline for the segmentation.

While we are on the topic of telephone calls, I do see some multi-channel support in this codebase. Can you explain (or is there a document) how lhotse would handle a dual channel phone call?

**Reply by pzelasko** [reply]
In general you would create a Recording that has an AudioSource pointing to a 2-channel file (e.g. stereo wav/sph) or two AudioSources pointing to separate single channel files. Check the Switchboard recipe for the former, and the AMI recipe for the latter. Then, supervisions also have a field indicating to which channel they correspond (one or more channels for each of them). So you can have supervision with channel=0 to indicate e.g. transcript from the left channel, and channel=1 for transcript of the right channel. 

When you create cuts, I am not 100% sure after recent changes from @desh2608, but I think you have the option to create either MultiCuts that can access both channels data or MonoCuts that correspond only to one channel. I think it depends on how you set the channels in the supervisions (e.g. for telephone speech, it makes sense that a supervision is on a single channel only; if you have AB or XY microphone setup, then you would probably have supervision on both channels; etc.).
